test/e2e/storage/testsuites/subpath.go

```diff
@@ -71,345 +72,351 @@ func (s *subPathTestSuite) getTestSuiteInfo() TestSuiteInfo {
+	// registers its own BeforeEach which creates the namespace. Beware that it
+	// also registers an AfterEach which renders f unusable. Any code using
+	// f must run inside an It or Context callback.
+	f := framework.NewDefaultFramework("provisioning")
 
-		AfterEach(func() {
-			if needsCleanup {
-				resource.cleanupResource(driver, pattern)
-			}
-		})
+	init := func() {
+		cs = f.ClientSet // needed for cleanup, f.ClientSet itself gets reset too early
 
-		testSubPath(&input)
-	})
-}
+		// Now do the more expensive test initialization.
+		config, testCleanup = driver.PrepareTest(f)
+		fsType := pattern.FsType
+		volType := pattern.VolType
 
-type subPathTestResource struct {
-	genericVolumeTestResource
+		resource = createGenericVolumeTestResource(driver, config, pattern)
 
-	roVolSource *v1.VolumeSource
-	pod         *v1.Pod
-	formatPod   *v1.Pod
-}
+		// Setup subPath test dependent resource
+		switch volType {
+		case testpatterns.InlineVolume:
+			if iDriver, ok := driver.(InlineVolumeTestDriver); ok {
+				roVolSource = iDriver.GetVolumeSource(true, fsType, resource.volume)
+			}
+		case testpatterns.PreprovisionedPV:
+			roVolSource = &v1.VolumeSource{
+				PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+					ClaimName: resource.pvc.Name,
+					ReadOnly:  true,
+				},
+			}
+		case testpatterns.DynamicPV:
+			roVolSource = &v1.VolumeSource{
+				PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+					ClaimName: resource.pvc.Name,
+					ReadOnly:  true,
+				},
+			}
+		default:
+			framework.Failf("SubPath test doesn't support: %s", volType)
+		}
 
-var _ TestResource = &subPathTestResource{}
+		subPath := f.Namespace.Name
+		pod = SubpathTestPod(f, subPath, resource.volType, resource.volSource, true)
+		pod.Spec.NodeName = config.ClientNodeName
+		pod.Spec.NodeSelector = config.ClientNodeSelector
 
-func (s *subPathTestResource) setupResource(driver TestDriver, pattern testpatterns.TestPattern) {
-	s.driver = driver
-	dInfo := s.driver.GetDriverInfo()
-	f := dInfo.Config.Framework
-	fsType := pattern.FsType
-	volType := pattern.VolType
+		formatPod = volumeFormatPod(f, resource.volSource)
+		formatPod.Spec.NodeName = config.ClientNodeName
+		formatPod.Spec.NodeSelector = config.ClientNodeSelector
 
-	// Setup generic test resource
-	s.genericVolumeTestResource.setupResource(driver, pattern)
+		subPathDir = filepath.Join(volumePath, subPath)
+		filePathInSubpath = filepath.Join(volumePath, fileName)
+		filePathInVolume = filepath.Join(subPathDir, fileName)
+	}
 
-	// Setup subPath test dependent resource
-	switch volType {
-	case testpatterns.InlineVolume:
-		if iDriver, ok := driver.(InlineVolumeTestDriver); ok {
-			s.roVolSource = iDriver.GetVolumeSource(true, fsType, s.genericVolumeTestResource.driverTestResource)
+	cleanup := func() {
+		if pod != nil {
+			By("Deleting pod")
+			err := framework.DeletePodWithWait(f, cs, pod)
+			Expect(err).ToNot(HaveOccurred(), "while deleting pod")
+			pod = nil
```

Comment:
this is suspicious to me. I want to avoid a pattern where we are reusing structs and just clearing field. Is there a way we can start from scratch instead?

Reply:
The test does start from scratch, with a fresh copy of all local variables. Setting the pointers to nil is meant to ensure that `cleanup` can be called more than once, should the need ever arise. Currently this isn't done, so I could remove the `pod = nil` and leave the pointers to already deleted items, but personally would prefer to keep this extra precaution.

Reply:
Hmm, I need to be more precise: each test starts with a fresh copy of the the local variables *if* `init` initializes all of them. I see how that might be something that can be missed for a complex test suite where some variables are only set under certain conditions.

This can be avoided by storing all these local variables in a struct and re-initializing the struct before setting individual members. I've done that in commit ec3655a1d40ced6b1873e627b736aae1cf242477.

I'm undecided whether it makes the code easier to read ("shared" variables clearly marked with `l.` prefix) or harder to read (extra prefix). @davidz627 Please let me know whether we should do it this way or by carefully setting all variables in init.